Load more data types

JoaoPedroDinis opened the pull request:
This PR fixes load of some data types that were having trouble loading like reading a bunch of datas, accounting, currency, scientific notation, phone number (still needs formatting to be done on string loaded), reads "TRUE" as true boolean value, fractions and percentages.

In sum this commit allow read of multiple data types but most importantly those available from the drop down box and reads results of formulas if present.

If anyone needs I can try to clarify this a bit more 👍 

Ps: #3 The code in there turned out to not fix anything, however the good news is that from my tests it reads and writes past Z column. 

So far for writing I can only see a few issues and that is, when writing it will write everything as general even if you set the data type for cell.DataType and if you load in excel 2019 it will try to auto repair and claim its corrupt, repairing shows all the data saved correctly minus the previously mentioned general data type for all cells.

JDinis committed:
Updated Project and Packages

Updated packages on both projects and updated the net standard framework to be able to use C# 8.

JDinis committed:
Copied xlsx and added a new model

Copied the existing xlsx and added more fake data to it

Created a copy of the existing model with extra fields

Created a test file and a test for xlsx reads

JDinis committed:
Updated TryParseData function

Down the road of my tests I've noticed that the problems of converting data on some formats was due to the usage of a '.' while C# expected ','

This behaviour might change from culture to culture.

Worksheet and the other vars were then moved outside TryParseDate as I noticed that if there is no DataType then we must guide ourselves by formatIds value.

JDinis committed:
Updated GetCellValue

Updated this function to be able to return an object so that less casts and data type convertions are needed.

Now I've added a switch to control data parses based on formatId if the cell doesn't contain DataType.

For more info on how to parse data from these formatIds I've looked into OpenXML docs but being too extensive I couldn't find much, however Microsoft does provide documentation on OpenXML and it can be found at: https://docs.microsoft.com/en-us/dotnet/api/documentformat.openxml.spreadsheet.numberingformat?view=openxml-2.8.1

Finally I've updated the return value of the Boolean case to return a true of false and thus avoiding parsing to string and viceversa to read and save boolean values.

Depending on wether cell.CellFormula is null (meaning it has the name implies that a formula exists) or not we send either cell.CellValue.Text or cell.InnerText

The 1st contains the result of the formula execution and the former contains the value we want if the cell value is static

JDinis committed:
Added more cases to GetCellValues

Added more cases and a new field to model with decimal data type to load accounting data

Updated the xlsx to add Debt header and values with accounting as data type

ghost commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/OfficeDev/Excel-IO?pullRequest=4) <br/>All CLA requirements met.

JoaoPedroDinis commented:
Forgot to mention I've updated it to net standard 2.1 and its nuget packages.

JoaoPedroDinis committed:
Adding forgotten tiny refactoring

This was the reason I updated to net standard 2.1 in the first place :P

Well that and C# 8

craigomatic commented on src/Excel.IO/ExcelExtensions.cs:
```diff
@@ -26,48 +26,144 @@ public static string GetColumn(this Cell cell)
-            }
-
-            return cell.CellReference.Value.Substring(0, endIndex);
+            return cell.CellReference?.Value?[0..^-1];
```
When I run any of the unit tests (including the one added as part of this PR) that does a read, it ends up with an ArgumentOutOfRangeException here.

JoaoPedroDinis replied:
Weird, I'll investigate and sorry for that :(

JoaoPedroDinis replied:
My bad I added an - and [0..^x] works like so: start from char 0 to last char - x where x is an int

JoaoPedroDinis replied:
All fixed up in this last push and again apologies because I added this over GitHub directly instead of thru ide so this last addition wasn't tested as it should've 😨

JoaoPedroDinis replied:
Turns out that CellReference is (at least in some cases, don't know if in all cases) Row so can't just subtract 1 char because when it reaches for example 10 we will have A10 and the code would've subtracted 1 char making it A1.

I've reverted back to the commit before tiny refactoring commit

JDinis committed:
Fix mistake of adding - to range

JDinis committed:
Rollback change to for loop

Reason: Found out CellReference appears to be Row and not Column so cant just subtract 1 char

JDinis committed:
Created Extension Method

Created extension method to replace either a dot or a comma with the current culture decimal separator char

craigomatic commented:
Sorry for slow follow up! 

Two things we should resolve:

1. I think we need to roll back to net standard 2.0 as 2.1 doesn't support .NET Framework
2. Can we merge the tests back into the one xlsx (use a 3rd sheet?) and have just the one cs file with tests?

JDinis committed:
Rollback to .Net Standard 2.0

JDinis committed:
Added new tests
Moved tests to ExcelConverterTests
Fixed some ids causing incorrect data parses

JoaoPedroDinis commented:
> Sorry for slow follow up!
> 
> Two things we should resolve:
> 
> 1. I think we need to roll back to net standard 2.0 as 2.1 doesn't support .NET Framework
> 2. Can we merge the tests back into the one xlsx (use a 3rd sheet?) and have just the one cs file with tests?

Hi again Craig,
Yes, it doesn't I hadn't seen this list before: https://docs.microsoft.com/en-us/dotnet/standard/net-standard
so I've rolled back

As for the tests being in another xlsx I didn't want to end up ruining data or something :), I've also done this and fixed some bugs found with new tests added.
